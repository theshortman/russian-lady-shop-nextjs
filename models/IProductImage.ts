export interface IProductImage {
  id: number;
  image_large: string;
  image_medium: string;
  image_small: string;
}
