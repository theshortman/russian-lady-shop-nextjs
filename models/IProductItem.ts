export interface IProductItem {
  id: number;
  size: number;
  quantity: number;
}
