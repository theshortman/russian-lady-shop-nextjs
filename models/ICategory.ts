import { IProduct } from "@/models/IProduct";

export interface ICategory {
  id: number;
  name: string;
  description: string;
  slug: string;
  products: IProduct[];
  prev_page_number: number | null;
  next_page_number: number | null;
}
