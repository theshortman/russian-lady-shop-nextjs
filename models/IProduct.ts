import { IProductImage } from "@/models/IProductImage";
import { IProductItem } from "@/models/IProductItem";

export interface IProduct {
  id: number;
  name: string;
  slug: string;
  description: string;
  detail: string;
  price: number;
  discount: number;
  new_price: number;
  product_images: IProductImage[];
  product_items: IProductItem[];
}
