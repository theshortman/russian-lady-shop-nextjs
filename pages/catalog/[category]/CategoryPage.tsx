import { ICategory } from "@/models/ICategory";
import type { NextPage } from "next";
import Head from "next/head";
import ProductItem from "./ProductItem";
import Pagination from "@/components/Pagination";

interface CategoryPageProps {
  category: ICategory;
}

const generatePageUrl = (categorySlug: string, page: number): string | null =>
  page ? `/catalog/${categorySlug}/page/${page}` : null;

const CategoryPage: NextPage<CategoryPageProps> = ({ category }) => {
  return (
    <>
      <Head>
        <title>
          {category.name} | {process.env.title}
        </title>
        <meta name="description" content={category.description} />
      </Head>
      <div>
        <h1>{category.name}</h1>
        <div className="row">
          {category.products.length ? (
            category.products.map((prod) => (
              <ProductItem
                categorySlug={category.slug}
                key={prod.id}
                product={prod}
              />
            ))
          ) : (
            <p>Ничего не найдено</p>
          )}
        </div>
        {(category.prev_page_number || category.next_page_number) && (
          <Pagination
            prevPageLink={generatePageUrl(
              category.slug,
              category.prev_page_number
            )}
            nextPageLink={generatePageUrl(
              category.slug,
              category.next_page_number
            )}
          />
        )}
      </div>
    </>
  );
};

export default CategoryPage;
