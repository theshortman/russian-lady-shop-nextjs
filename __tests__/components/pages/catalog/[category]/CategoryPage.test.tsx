import { render } from "@testing-library/react";
import CategoryPage from "@/components/pages/catalog/[category]/CategoryPage";
import { ICategory } from "@/models/ICategory";
import { IProduct } from "@/models/IProduct";

describe("Category page", () => {
  const product: IProduct = {
    id: 1,
    name: "Test product name",
    slug: "test-product-slug",
    description: "Test product description",
    detail: "test product detail",
    price: 1000,
    discount: 10,
    new_price: 900,
    product_images: [
      {
        id: 1,
        image_large: "first-image-large",
        image_medium: "first-image-medium",
        image_small: "first-image-small",
      },
      {
        id: 2,
        image_large: "second-image-large",
        image_medium: "second-image-medium",
        image_small: "second-image-small",
      },
    ],
    product_items: [
      {
        id: 1,
        size: 50,
        quantity: 1,
      },
    ],
  };

  const category: ICategory = {
    id: 4,
    name: "Test category name",
    description: "Test category description",
    slug: "test-category-slug",
    products: [],
    prev_page_number: null,
    next_page_number: null,
  };

  beforeEach(() => {
    category.products = [];
    category.prev_page_number = null;
    category.next_page_number = null;
  });

  test("renders CategoryPage unchanged if products do not exist", () => {
    const { container } = render(<CategoryPage category={category} />);
    expect(container).toMatchSnapshot();
  });

  test("renders CategoryPage unchanged if products exist", () => {
    category.products.push(product);
    const { container } = render(<CategoryPage category={category} />);
    expect(container).toMatchSnapshot();
  });

  test("renders CategoryPage unchanged if there is a prev page", () => {
    category.prev_page_number = 1;
    const { container } = render(<CategoryPage category={category} />);
    expect(container).toMatchSnapshot();
  });

  test("renders CategoryPage unchanged if there is a next page", () => {
    category.next_page_number = 2;
    const { container } = render(<CategoryPage category={category} />);
    expect(container).toMatchSnapshot();
  });

  test("renders CategoryPage unchanged if there is a prev and next page", () => {
    category.next_page_number = 1;
    category.next_page_number = 2;
    const { container } = render(<CategoryPage category={category} />);
    expect(container).toMatchSnapshot();
  });
});
