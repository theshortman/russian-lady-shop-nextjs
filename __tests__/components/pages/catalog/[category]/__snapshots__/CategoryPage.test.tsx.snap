// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Category page renders CategoryPage unchanged if products do not exist 1`] = `
<div>
  <div>
    <h1>
      Test category name
    </h1>
    <div
      class="row"
    >
      <p>
        Ничего не найдено
      </p>
    </div>
  </div>
</div>
`;

exports[`Category page renders CategoryPage unchanged if products exist 1`] = `
<div>
  <div>
    <h1>
      Test category name
    </h1>
    <div
      class="row"
    >
      <div
        class="col-sm-6 col-md-6 col-lg-4 mb-3"
      >
        <div
          class="row"
        >
          <div
            class="col-7 col-sm-8 col-md-8"
          >
            <a
              href="/catalog/test-category-slug/test-product-slug"
            >
              <svg
                aria-label="No photo"
                class="img-fluid img-thumbnail"
                focusable="false"
                height="912"
                preserveAspectRatio="xMidYMid slice"
                role="img"
                width="600"
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  fill="#ffffff"
                  height="100%"
                  width="100%"
                />
                <text
                  dy=".3em"
                  fill="#dee2e6"
                  text-anchor="middle"
                  x="50%"
                  y="50%"
                >
                  Loading...
                </text>
              </svg>
              <img
                alt="Test product name"
                class="img-fluid img-thumbnail d-none"
                data-testid="product-img"
                src="first-image-medium"
              />
            </a>
          </div>
          <div
            class="col-5 col-sm-4 col-md-4 ps-0"
          >
            <div
              class="mb-1"
            >
              <div>
                <span
                  class="me-1 badge bg-success"
                >
                  900
                  RUB
                </span>
                <div
                  class="badge bg-danger"
                >
                  -
                  10
                  %
                </div>
              </div>
            </div>
            <div
              class="mb-2"
            >
              <div>
                <span
                  class="badge bg-secondary me-1"
                >
                  50
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Category page renders CategoryPage unchanged if there is a next page 1`] = `
<div>
  <div>
    <h1>
      Test category name
    </h1>
    <div
      class="row"
    >
      <p>
        Ничего не найдено
      </p>
    </div>
    <nav
      aria-label="Page navigation"
    >
      <ul
        class="pagination justify-content-center"
      >
        <li
          class="page-item"
        >
          <a
            aria-label="Next"
            class="page-link"
            href="/catalog/test-category-slug/page/2"
          >
            След.
          </a>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;

exports[`Category page renders CategoryPage unchanged if there is a prev and next page 1`] = `
<div>
  <div>
    <h1>
      Test category name
    </h1>
    <div
      class="row"
    >
      <p>
        Ничего не найдено
      </p>
    </div>
    <nav
      aria-label="Page navigation"
    >
      <ul
        class="pagination justify-content-center"
      >
        <li
          class="page-item"
        >
          <a
            aria-label="Next"
            class="page-link"
            href="/catalog/test-category-slug/page/2"
          >
            След.
          </a>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;

exports[`Category page renders CategoryPage unchanged if there is a prev page 1`] = `
<div>
  <div>
    <h1>
      Test category name
    </h1>
    <div
      class="row"
    >
      <p>
        Ничего не найдено
      </p>
    </div>
    <nav
      aria-label="Page navigation"
    >
      <ul
        class="pagination justify-content-center"
      >
        <li
          class="page-item"
        >
          <a
            aria-label="Previous"
            class="page-link"
            href="/catalog/test-category-slug/page/1"
          >
            Пред.
          </a>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;
